hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/FlinkRowDataReaderContext.java

```diff
@@ -0,0 +1,257 @@
+    RowDataFileReaderFactories.Factory readerFactory = RowDataFileReaderFactories.getFactory(HoodieFileFormat.PARQUET);
+    RowDataFileReader fileReader = readerFactory.createFileReader(schemaManager, getStorageConfiguration());
+
+    List<String> fieldNames = dataSchema.getFields().stream().map(Schema.Field::name).collect(Collectors.toList());
+    List<DataType> fieldTypes = dataSchema.getFields().stream().map(
+        f -> AvroSchemaConverter.convertToDataType(f.schema())).collect(Collectors.toList());
+    int[] selectedFields = requiredSchema.getFields().stream().map(Schema.Field::name)
+        .map(fieldNames::indexOf)
+        .mapToInt(i -> i)
+        .toArray();
+
+    return fileReader.getRowDataIterator(
+        fieldNames, fieldTypes, selectedFields, predicates, filePath, start, length);
+  }
+
+  @Override
+  public Option<HoodieRecordMerger> getRecordMerger(RecordMergeMode mergeMode, String mergeStrategyId, String mergeImplClasses) {
+    switch (mergeMode) {
+      case EVENT_TIME_ORDERING:
+        return Option.of(new EventTimeFlinkRecordMerger());
+      case COMMIT_TIME_ORDERING:
+        return Option.of(new CommitTimeFlinkRecordMerger());
+      default:
+        Option<HoodieRecordMerger> mergerClass =
+            HoodieRecordUtils.createValidRecordMerger(EngineType.FLINK, mergeImplClasses, mergeStrategyId);
+        if (mergerClass.isEmpty()) {
+          throw new HoodieValidationException("No valid flink merger implementation set for `"
+              + RECORD_MERGE_IMPL_CLASSES_WRITE_CONFIG_KEY + "`");
+        }
+        return mergerClass;
+    }
+  }
+
+  @Override
+  public Object getValue(RowData record, Schema schema, String fieldName) {
+    RowData.FieldGetter fieldGetter = RowDataAvroQueryContexts.fromAvroSchema(schema).getFieldQueryContext(fieldName).getFieldGetter();
+    return fieldGetter.getFieldOrNull(record);
+  }
+
+  @Override
+  public String getRecordKey(RowData record, Schema schema) {
+    return Objects.toString(getValue(record, schema, RECORD_KEY_METADATA_FIELD));
+  }
+
+  @Override
+  public HoodieRecord<RowData> constructHoodieRecord(Option<RowData> recordOption, Map<String, Object> metadataMap) {
+    HoodieKey hoodieKey = new HoodieKey(
+        (String) metadataMap.get(INTERNAL_META_RECORD_KEY),
+        (String) metadataMap.get(INTERNAL_META_PARTITION_PATH));
+    RowData rowData = recordOption.get();
+    // delete record
+    if (recordOption.isEmpty()) {
+      Comparable orderingValue;
+      if (metadataMap.containsKey(INTERNAL_META_ORDERING_FIELD)) {
+        orderingValue = (Comparable) metadataMap.get(INTERNAL_META_ORDERING_FIELD);
+      } else {
+        throw new HoodieException("There should be ordering value in metadataMap.");
+      }
+      return new HoodieEmptyRecord<>(hoodieKey, HoodieOperation.DELETE, orderingValue, HoodieRecord.HoodieRecordType.FLINK);
+    }
+    return new HoodieFlinkRecord(hoodieKey, rowData);
```

Comment:
why not also set up ordering value here?

Reply:
because there exists ordering_value in `metadataMap` only when the record is delete record, see `HoodieReaderContext#generateMetadataForRecord`. For insert/upsert records, `HoodieFlinkRecord` is constructed without ordering value, which will be extracted lazily if necessary.